web/static/js/components/thesis-editor.js

```diff
@@ -223,6 +264,44 @@ class ThesisEditor extends React.Component {
     }
   }
 
+  importContent (t, d, page) {
+    switch (t) {
+      case 'image':
+      case 'background_image':
+        return this.imageEditor.uploadAndSet(d, page, () => { this.updateImportedContentCompletedCount() })
+      case 'text':
+        this.updateImportedContentCompletedCount()
+        return this.textEditor.set(d.name, d)
+      case 'html':
+        this.updateImportedContentCompletedCount()
+        return this.htmlEditor.set(d.name, d)
+      case 'raw_html':
+        this.updateImportedContentCompletedCount()
+        return this.rawHtmlEditor.set(d.name, d)
+      default:
+        return true
+    }
+  }
+
+  updateImportedContentCompletedCount () {
+    this.setState({importContentCompletedCount: ++this.state.importContentCompletedCount}, () => {
```

Comment:
*Important:* This is a dangerous `setState` call that relies on potentially stale state, and you're mutating state with that `++` call beside. You'll want to pass through a function here before setting state.

```javascript
this.setState((prevState, props) => {
  const newCompletedCount = prevState.importContentCompletedCount + 1
  if (newCompletedCount >== this.state.importContentQueueCount) {
    clearTimeout(window.importContentCounterTimeout)

    return {
      importContentQueueCount: 0,
      importContentCompletedCount: 0,
      importProgress: null,
      trayOpen: false
    }
  } else {
    return { importContentCompletedCount: newCompletedCount }
  }
})
```